src/oscar/apps/basket/views.py

```diff
@@ -332,6 +333,13 @@ def form_invalid(self, form):
         clean_msgs = [m.replace('* ', '') for m in msgs if m.startswith('* ')]
         messages.error(self.request, ",".join(clean_msgs))
 
+        # We serialize the POST data with JSONSerializer before adding it to the session.
+        # Without this, we could expose the site to a security vulnerability
+        # if the SESSION_SERIALIZER has been configured to 'django.contrib.sessions.serializers.PickleSerializer'.
+        # see: https://docs.djangoproject.com/en/3.2/topics/http/sessions/#cookie-session-backend
+        serialized_data = JSONSerializer().dumps(self.request.POST)
+        self.request.session["add_to_basket_form_post_data"] = serialized_data.decode("latin-1")
```

Comment:
I think this may be prone to race conditions because the session key is global. Imagine if two items are added to the basket at the same time (multiple tabs, multiple forms on a page) - it's possible for this response to end up going to the wrong form. Maybe we should suffix it with the product ID?

Reply:
Yes I think this is a good idea. I will change it.